ppb/engine.py

```diff
@@ -39,7 +46,170 @@ def _get_handler_name(txt):
     _get_handler_name(x)
 
 
-class GameEngine(LoggingMixin):
+class EngineChildren(Children):
+    """
+    Acts as a Children collection for engines:
+
+    * Scenes are managed in their own stack
+    * Systems have context managers
+    * Manipulating Scenes must be through pushing and popping
+    * Manipulating Systems is disallowed while the engine is running.
+    * Only the active (topmost) Scene is exposed as a child
+    * The iteration order is defined as: Systems, Current Scene, anything else
+    """
+    entered: bool
+
+    def __init__(self):
+        super().__init__()
+        self._scenes = []
+        self._systems = set()
+
+        self._stack = ExitStack()
+        self.entered = False
+
+    def __contains__(self, item: Hashable) -> bool:
+        return (
+            item in self._all or
+            item in self._scenes or
+            item in self._systems
+        )
+
+    def __iter__(self) -> Iterator[Hashable]:
+        yield from self._systems
+        if self._scenes:
+            yield self._scenes[-1]
+        yield from self._all
+
+    def __len__(self) -> int:
+        return len(self._all)
+
+    @property
+    def current_scene(self):
+        """
+        The top of the scene stack.
+
+        :return: The currently running scene.
+        :rtype: ppb.BaseScene
+        """
+        try:
+            return self._scenes[-1]
+        except IndexError:
+            return None
+
+    def add(self, child: Hashable, tags: Iterable[Hashable] = ()) -> Hashable:
+        """
+        Add a child.
+
+        :param child: Any Hashable object. The item to be added.
+        :param tags: An iterable of Hashable objects. Values that can be used to
+              retrieve a group containing the child.
+
+        Note that Scenes and Systems have special restrictions.
+
+        Examples: ::
+
+            children.add(MyObject())
+
+            children.add(MyObject(), tags=("red", "blue")
+        """
+        # Ugh, this is a copy of the implementation in Children.
```

Comment:
Not for this PR but we could consider a "validate_child" helper function in `Children` that can be called in implementations of `add` and overridden for things like the check on line 122.

Need to consider the implementation considering we don't want to rewrite 131-134 either.